generator/config/query/regex.yaml

```diff
@@ -11,3 +11,27 @@ arguments:
         name: regex
         type:
             - regex
+
+tests:
+    -
+        name: 'Perform a LIKE Match'
+        link: 'https://www.mongodb.com/docs/manual/reference/operator/query/regex/#perform-a-like-match'
+        pipeline:
+            -
+                $match:
+                    sku:
+                        # Should be nested in $regex
+                        $regularExpression:
+                            pattern: '789$'
+                            options: ''
```

Comment:
`$regex` is a special key. I can't use it for a regular object.

```php
$json = <<<'JSON'
{
    "foo": {
        "$regex": {
            "$regularExpression": {
                "pattern": "789$",
                "options": "i"
            }
        }
    }
}
JSON;

$document = fromJSON($json);
```

Reply:
That exception originates from [`_bson_json_read_start_map()`](https://github.com/mongodb/mongo-c-driver/blob/1.24.4/src/libbson/src/bson/bson-json.c#L1215)

Also discussed in the Extended JSON spec:

 * [Parsers](https://github.com/mongodb/specifications/blob/master/source/extended-json.rst#parsers)
 * [Of Canonical Type Wrapper Formats: RegularExpression](https://github.com/mongodb/specifications/blob/master/source/extended-json.rst#regularexpression)
 * [Q&A](https://github.com/mongodb/specifications/blob/master/source/extended-json.rst#q-a)

If you're using `$regex` as a query operator, the "Parsers" section above says:

> A parser that accepts Legacy Extended JSON MUST be configurable such that a JSON text of a MongoDB query filter containing the [regex](https://www.mongodb.com/docs/manual/reference/operator/query/regex/) query operator can be parsed.

And then it proceeds to give two examples. The first example resembles what you're trying to do here, where the `$regex` value is the extended JSON representation of a regex object (using the `$regularExpression` key). I'm not sure why that's prohibited in libbson, but it may be worth asking the C team about.

That aside, is anything preventing you from using the following?

```
{ $regex: 'pattern', $options: '<options>' }
```

Reply:
> That aside, is anything preventing you from using the following?
> `{ $regex: 'pattern', $options: '<options>' }`

This will result in `new MongoDB\BSON\Regex('pattern', '<options'>)`.

The codec wraps the regex in the `$regex` operator for consistency with other operators.

The expected query that I need to encode in Yaml/Json is:
```
(object) ['sku' => (object) ['$regex' => new Regex('789$', '')]]
```

Reply:
```php
$query = (object) ['sku' => (object) ['$regex' => new MongoDB\BSON\Regex('789$', '')]];

MongoDB\BSON\fromJSON(
    MongoDB\BSON\toJson(
        MongoDB\BSON\fromPHP($x)
    )
);

// Warning: Uncaught MongoDB\Driver\Exception\UnexpectedValueException: Unexpected nested object value for "$regex" key in php shell code:1
```

Reply:
Issue solved using `MongoDB\BSON\Document`: https://github.com/mongodb/mongo-php-builder/pull/13

```php
$query = (object) ['sku' => (object) ['$regex' => new \MongoDB\BSON\Regex('789$', '')]];

$json = MongoDB\BSON\Document::fromPHP($query)->toCanonicalExtendedJSON();
```